watcher.go

```diff
@@ -409,101 +356,256 @@ func (w *Watcher) Stop() {
-		delete(t.deps[d], tmplID)
-		delete(t.tpls[tmplID], d)
+	for id, view := range t.views {
+		delete(t.views, id)
+		if view == nil {
+			continue
+		}
+		view.stop()
+	}
+}
+
+// Return all views for a notifier
+func (t *tracker) notifiersFor(v *view) []Notifier {
+	viewID := v.ID()
+	results := make([]Notifier, 0, 8)
+	for _, tp := range t.tracked {
+		if tp.view == viewID {
+			results = append(results, t.notifiers[tp.notify])
+		}
 	}
+	return results
+}
+
+// initialized returns true if the view has had its data fetched at least once
+func (t *tracker) initialized(viewID string) bool {
+	return t.views[viewID].receivedData
 }
 
-func (t *tracker) findUnused(depIDs []string) map[string]struct{} {
-	t.RLock()
-	defer t.RUnlock()
-	result := make(map[string]struct{})
-	for _, depID := range depIDs {
-		if len(t.deps[depID]) == 0 {
-			result[depID] = struct{}{}
+// complete returns true if every dependency used has been initialized
+// ie. it returns true if all values have been fetched
+func (t *tracker) complete(n Notifier) bool {
+	for _, tp := range t.tracked {
+		thisNotifier := tp.notify == n.ID()
+		//		fmt.Println(thisNotifier, tp.inUse, !t.initialized(tp.view))
+		//		fmt.Printf("%#v\n", tp)
+		if thisNotifier && tp.inUse && !t.initialized(tp.view) {
+			return false
 		}
 	}
-	return result
+	return true
 }
 
-func (t *tracker) templateDeps(tmplID string) (map[string]struct{}, bool) {
-	t.RLock()
-	defer t.RUnlock()
-	deps, ok := t.tpls[tmplID]
-	return deps, ok
+// Clean out un-used trackedPair entries, views and notifiers
+// Checks based on passed in notifier, ignores others.
+func (t *tracker) sweep(n Notifier) {
+	t.Lock()
+	defer t.Unlock()
+	used := make(map[string]struct{})
+	// remove tracked that were used
+	tmp := t.tracked[:0]
+	for _, tp := range t.tracked {
+		otherNotifier := tp.notify != n.ID()
+		if tp.inUse || otherNotifier {
+			tmp = append(tmp, tp.refresh())
+			used[tp.view] = struct{}{}
+			used[tp.notify] = struct{}{}
+		}
+	}
+	t.tracked = tmp
+	// remove views/notifiers no longer referenced
+	for v := range t.views {
+		if _, ok := used[v]; !ok {
+			delete(t.views, v)
+		}
+	}
+	for n := range t.notifiers {
+		if _, ok := used[n]; !ok {
+			delete(t.views, n)
+		}
+	}
+}
+
+// dummy Notifier for use by vault token above and in tests
+type dummyNotifier struct {
+	name string
+	deps chan dep.Dependency
+}
+
+func fakeNotifier(name string) *dummyNotifier {
+	return &dummyNotifier{name: name, deps: make(chan dep.Dependency, 100)}
+}
+func (n *dummyNotifier) Notify(d dep.Dependency) {
+	n.deps <- d
+}
+func (n *dummyNotifier) ID() string {
+	return string(n.name)
+}
+func (n *dummyNotifier) count() int {
+	return len(n.deps)
+}
+func (n *dummyNotifier) ids() []string {
+	results := make([]string, len(n.deps))
+	for i := 0; i < len(n.deps); i++ {
+		d := <-n.deps
+		results[i] = d.String()
+		n.deps <- d
+	}
+	return results
+}
+func (n *dummyNotifier) drain() {
+	for range n.deps {
+		// just draining
+	}
```

Comment:
What's the purpose of `drain()`? This blocks without a way to stopping the function without the caller closing the channel.

```golang
close(n.deps)
```

Reply:
Drain should have been removed. It isn't used and that method isn't even right.